addons/mail/models/mail_thread.py

```diff
@@ -2248,6 +2248,14 @@ def _notify_record_by_email(self, message, recipients_data, msg_vals=False,
         }
         base_mail_values = self._notify_by_email_add_values(base_mail_values)
 
+        # Clean the context to get rid of residual default_* keys that could cause issues afterward
+        # during the mail.mail and mail.notification generation.
+        # Example: 'default_state' would refer to the default state of a previously created record
+        # from another model that in turns triggers an assignation notification that ends up here.
+        # This will lead to a traceback when trying to create a mail.mail with this state value that
+        # doesn't exist.
+        self = self.with_context(clean_context(self._context))
```

Comment:
Shouldn't we limit the clean context at mail.mail and mail.notification creation ?

Reply:
You mean something like:

`email = Mail.with_context(clean_context(self._context)).create(create_values)` ?
And the same for notifications?

That is something I considered but I followed the example here:
https://github.com/odoo/odoo/blob/13.0/addons/mail/models/mail_thread.py#L595
Also was not sure if cleaning context in a loop would be a good idea.

But it does indeed seem safer so the global context in untouched when you leave this method.

Reply:
Was more thing of preparing
```
SafeMail = self.env['mail.mail'].with_context(clean_context)
SafeNotification = self.env['mail.notification'].with_context(clean_context)
```

Then using SafeMail / SafeNotification when necessary. That way we prepare cleaned environments and use it locally, instead of rewriting self and instead of using clean context too much times.